core/state_transition.go

```diff
@@ -373,232 +366,68 @@ func (st *StateTransition) StakingTransitionDb() (usedGas uint64, err error) {
-	}
-
-	if !CanTransfer(st.state, createValidator.ValidatorAddress, createValidator.Amount) {
-		return errInsufficientBalanceForStake
-	}
-
-	v, err := staking.CreateValidatorFromNewMsg(createValidator, blockNum)
+func (st *StateTransition) verifyAndApplyCreateValidatorTx(
+	createValidator *staking.CreateValidator, blockNum *big.Int,
+) error {
+	wrapper, err := VerifyAndCreateValidatorFromMsg(st.state, st.evm.EpochNumber, blockNum, createValidator)
 	if err != nil {
 		return err
 	}
-
-	zero := big.NewInt(0)
-	wrapper := staking.ValidatorWrapper{}
-	wrapper.Validator = *v
-	wrapper.Delegations = []staking.Delegation{
-		staking.NewDelegation(v.Address, createValidator.Amount),
-	}
-	wrapper.Snapshot.Epoch = st.evm.EpochNumber
-	wrapper.Snapshot.NumBlocksSigned = zero
-	wrapper.Snapshot.NumBlocksToSign = zero
-
-	if err := st.state.UpdateStakingInfo(v.Address, &wrapper); err != nil {
+	if err := st.state.UpdateStakingInfo(wrapper.Validator.Address, wrapper); err != nil {
 		return err
 	}
-
-	st.state.SetValidatorFlag(v.Address)
-	st.state.SubBalance(v.Address, createValidator.Amount)
+	st.state.SetValidatorFlag(wrapper.Validator.Address)
+	st.state.SubBalance(wrapper.Address, createValidator.Amount)
 	return nil
 }
 
-func (st *StateTransition) applyEditValidatorTx(
+func (st *StateTransition) verifyAndApplyEditValidatorTx(
 	editValidator *staking.EditValidator, blockNum *big.Int,
 ) error {
-	if !st.state.IsValidator(editValidator.ValidatorAddress) {
-		return errValidatorNotExist
-	}
-
-	wrapper := st.state.GetStakingInfo(editValidator.ValidatorAddress)
-	if wrapper == nil {
-		return errValidatorNotExist
-	}
-
-	if err := staking.UpdateValidatorFromEditMsg(&wrapper.Validator, editValidator); err != nil {
+	wrapper, err := VerifyAndEditValidatorFromMsg(st.state, st.bc, blockNum, editValidator)
+	if err != nil {
 		return err
 	}
-	newRate := wrapper.Validator.Rate
+	return st.state.UpdateStakingInfo(wrapper.Address, wrapper)
+}
 
-	// TODO: make sure we are reading from the correct snapshot
-	snapshotValidator, err := st.bc.ReadValidatorSnapshot(wrapper.Address)
+func (st *StateTransition) verifyAndApplyDelegateTx(delegate *staking.Delegate) error {
+	wrapper, err := VerifyAndDelegateFromMsg(st.state, delegate)
 	if err != nil {
 		return err
 	}
-	rateAtBeginningOfEpoch := snapshotValidator.Validator.Rate
-
-	if rateAtBeginningOfEpoch.IsNil() || (!newRate.IsNil() && !rateAtBeginningOfEpoch.Equal(newRate)) {
-		wrapper.Validator.UpdateHeight = blockNum
-	}
-
-	if newRate.Sub(rateAtBeginningOfEpoch).Abs().GT(wrapper.Validator.MaxChangeRate) {
-		return errCommissionRateChangeTooFast
-	}
-
-	if newRate.GT(wrapper.Validator.MaxRate) {
-		return errCommissionRateChangeTooHigh
-	}
-
-	if err := st.state.UpdateStakingInfo(wrapper.Address, wrapper); err != nil {
+	if err := st.state.UpdateStakingInfo(wrapper.Validator.Address, wrapper); err != nil {
 		return err
 	}
+	st.state.SubBalance(delegate.DelegatorAddress, delegate.Amount)
```

Comment:
This shouldn't be reducing by the whole delegate.Amount. The logic is this: to get the token for delegation, we first go into the list of undelegated tokens and use them first. If undelegated token are all used for delegation, then we use the account balance for the rest. So here it should be whatever is left in the "delegateBalance" in the original logic.

Reply:
Oh I see, okay will return the balance to be deducted from the delegator in `VerifyAndDelegateFromMsg`